radae/dsp.py

```diff
@@ -52,19 +52,55 @@ def __init__(self, Ntap, Fs_Hz, bandwidth_Hz, centre_freq_Hz):
       assert np.all(self.h == np.flip(self.h))
 
       self.mem = np.zeros(self.Ntap-1, dtype=np.csingle)
+
+      # Perform initial allocation of x_mem for the BPF.
+      self.x_mem = np.zeros(self.Ntap-1, dtype=np.csingle)
+
+      # Set initial value for array length. If we get a larger input,
+      # this is what will trigger reallocation.
+      self.n = -1
+
       self.phase = 1 + 0j
 
    def bpf(self, x):
       n = len(x)
       phase_vec = self.phase*np.exp(-1j*self.alpha*np.arange(1,n+1))
       x_baseband = x*phase_vec                                         # mix down to baseband
-      x_mem = np.concatenate([self.mem,x_baseband])                    # pre-pend filter memory
-      x_filt = np.zeros(n, dtype=np.csingle)
-      for i in np.arange(n):
-         x_filt[i] = np.dot(x_mem[i:i+self.Ntap],self.h)
-      self.mem = x_mem[-self.Ntap-1:]                                  # save filter state for next time
+
+      # Reallocate filter memory if the input is larger than the previous call.
+      if n > self.n:
+          self.x_filt = np.zeros(n, dtype=np.csingle)
+          self.n = n
+
+      # Reallocate x_mem if x_baseband size changes
+      if len(self.x_mem) != (len(self.mem) + len(x_baseband)):
+          self.x_mem = np.zeros(len(self.mem) + len(x_baseband), dtype=np.csingle)
```

Comment:
I like the idea of pre-allocating `x_filt`, however lets do it once at init time, based on the maximum buffer length pased into the init function.  Have an assert at run time to make sure we don't get passed a bigger `n` than we have room for.

Reply:
L75 performs a run time allocation of self.x_mem?  The assert should check the length of x_mem.